bitcoin/exchange/bitcoinprices.go

```diff
@@ -74,151 +92,103 @@ func (b *BitcoinPriceFetcher) UnitsPerCoin() int {
 	return SatoshiPerBTC
 }
 
-func (b *BitcoinPriceFetcher) run() {
-	b.fetchCurrentRates()
-	ticker := time.NewTicker(time.Minute * 15)
-	for range ticker.C {
-		b.fetchCurrentRates()
-	}
-}
-
 func (b *BitcoinPriceFetcher) fetchCurrentRates() error {
 	b.Lock()
 	defer b.Unlock()
 	for _, provider := range b.providers {
 		err := provider.fetch()
 		if err == nil {
+			log.Notice("Fetched current Bitcoin rates from " + provider.fetchUrl)
 			return nil
 		}
 	}
 	log.Error("Failed to fetch bitcoin exchange rates")
 	return errors.New("All exchange rate API queries failed")
 }
 
-type BitcoinAverage struct {
-	cache  map[string]float64
-	client *http.Client
-}
-
-func (b *BitcoinAverage) fetch() (err error) {
-	resp, err := b.client.Get("https://ticker.openbazaar.org/api")
+func (provider *ExchangeRateProvider) fetch() (err error) {
+	if len(provider.fetchUrl) == 0 {
+		err = errors.New("Provider has no fetchUrl")
+		return err
+	}
+	log.Info("Fetching rates at " + provider.fetchUrl)
+	resp, err := provider.client.Get(provider.fetchUrl)
 	if err != nil {
+		log.Error("Failed to fetch from "+provider.fetchUrl, err)
 		return err
 	}
-	return b.decode(resp.Body)
-}
-
-func (b *BitcoinAverage) decode(body io.ReadCloser) (err error) {
-	decoder := json.NewDecoder(body)
-	var data map[string]interface{}
-	err = decoder.Decode(&data)
+	decoder := json.NewDecoder(resp.Body)
+	var dataMap interface{}
+	err = decoder.Decode(&dataMap)
 	if err != nil {
+		log.Error("Failed to decode JSON from "+provider.fetchUrl, err)
 		return err
 	}
+	return provider.decoder.decode(dataMap, provider.cache)
+}
+
+func (b *BitcoinPriceFetcher) run() {
+	b.fetchCurrentRates()
+	ticker := time.NewTicker(time.Minute * 15)
+	for range ticker.C {
+		b.fetchCurrentRates()
+	}
+}
+
+// Decoders
+func (b BitcoinAverageDecoder) decode(dat interface{}, cache map[string]float64) (err error) {
+	data := dat.(map[string]interface{})
 	for k, v := range data {
 		if k != "timestamp" {
 			val, ok := v.(map[string]interface{})
 			if !ok {
-				return errors.New("Type assertion failed")
+				return errors.New(reflect.TypeOf(b).Name() + ".decode: Type assertion failed")
 			}
 			price, ok := val["last"].(float64)
 			if !ok {
-				return errors.New("Type assertion failed")
+				return errors.New(reflect.TypeOf(b).Name() + ".decode: Type assertion failed, missing 'last' (float) field")
```

Comment:
added more descriptive error description of what decoder is failing, and in some cases on what field it's failing.